substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/BufferNodeAccess.java

```diff
@@ -80,34 +71,47 @@ public static boolean tryLock(JfrBufferNode node) {
     }
 
     @Uninterruptible(reason = "Locking without transition requires that the whole critical section is uninterruptible.", callerMustBe = true)
-    public static void lockNoTransition(JfrBufferNode node) {
+    public static void lockNoTransition(BufferNode node) {
         assert node.isNonNull();
         NativeSpinLockUtils.lockNoTransition(ptrToLock(node));
         setLockOwner(node);
     }
 
     @Uninterruptible(reason = "Locking without transition requires that the whole critical section is uninterruptible.", callerMustBe = true)
-    public static void unlock(JfrBufferNode node) {
+    public static void unlock(BufferNode node) {
         assert node.isNonNull();
         assert isLockedByCurrentThread(node);
         node.setLockOwner(WordFactory.nullPointer());
         NativeSpinLockUtils.unlock(ptrToLock(node));
     }
 
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
-    public static boolean isLockedByCurrentThread(JfrBufferNode node) {
+    public static boolean isLockedByCurrentThread(BufferNode node) {
         assert CurrentIsolate.getCurrentThread().isNonNull();
         return node.isNonNull() && node.getLockOwner() == CurrentIsolate.getCurrentThread();
     }
 
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
-    private static void setLockOwner(JfrBufferNode node) {
+    private static void setLockOwner(BufferNode node) {
         assert node.getLockOwner().isNull();
         node.setLockOwner(CurrentIsolate.getCurrentThread());
     }
 
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
-    private static CIntPointer ptrToLock(JfrBufferNode node) {
-        return (CIntPointer) ((Pointer) node).add(JfrBufferNode.offsetOfLock());
+    protected static CIntPointer ptrToLock(BufferNode node) {
+        return (CIntPointer) ((Pointer) node).add(BufferNode.offsetOfLock());
+    }
+
+    /** Should be used instead of {@link com.oracle.svm.core.jfr.BufferNode#getBuffer}. */
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static JfrBuffer getJfrBuffer(BufferNode node) {
+        assert isLockedByCurrentThread(node) || VMOperation.isInProgressAtSafepoint();
+        return (JfrBuffer) node.getBuffer();
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static SamplerBuffer getSamplerBuffer(BufferNode node) {
```

Comment:
Just as a note: in the long run, we should probably try to get rid of this special buffer. This should be possible if we improve the `JfrBuffer`/`JfrBufferNode`/`JfrBufferList` infrastructure a bit.